OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -1005,7 +1010,10 @@ void BuildUnit(string category, bool buildRandom)
 				ChooseRandomUnitToBuild(queue) :
 				ChooseUnitToBuild(queue);
 
-			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name))
+			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name
+				&& (!Info.UnitLimits.ContainsKey(unit.Name)
+					|| World.Actors.Where(a => a.Owner == Player).Count(a => a.Info.Name == unit.Name) < Info.UnitLimits[unit.Name])))
```

Comment:
This will be quite bad for performance.
Consider counting and storing the number of units per type when they are built (would need to be updated when they are destroyed, which is the tricky part).